src/test/function_test/test_detect_hotspot.cpp

```diff
@@ -0,0 +1,340 @@
+        for (partition_index = 0; partition_index < partitions.size(); partition_index++) {
+            req.pid = dsn::gpid(app_id, partition_index);
+            auto errinfo =
+                ddl_client->detect_hotkey(partitions[partition_index].primary, req, resp);
+            ASSERT_EQ(errinfo, dsn::ERR_OK);
+            if (!resp.hotkey_result.empty()) {
+                find_hotkey = true;
+                break;
+            }
+        }
+        ASSERT_TRUE(find_hotkey);
+        ASSERT_EQ(resp.err, dsn::ERR_OK);
+        ASSERT_EQ(resp.hotkey_result, "ThisisahotkeyThisisahotkey");
+
+        // Wait for collector sending the next start detecting command
+        sleep(15);
+
+        req.action = dsn::replication::detect_action::STOP;
+        auto errinfo = ddl_client->detect_hotkey(partitions[partition_index].primary, req, resp);
+        ASSERT_EQ(errinfo, dsn::ERR_OK);
+        ASSERT_EQ(resp.err, dsn::ERR_OK);
+
+        req.action = dsn::replication::detect_action::QUERY;
+        for (partition_index = 0; partition_index < partitions.size(); partition_index++) {
+            req.pid = dsn::gpid(app_id, partition_index);
+            auto errinfo =
+                ddl_client->detect_hotkey(partitions[partition_index].primary, req, resp);
+            ASSERT_EQ(errinfo, dsn::ERR_OK);
+            ASSERT_EQ(resp.err_hint,
+                      "Can't get hotkey now, now state: hotkey_collector_state::STOPPED");
+        }
+    }
+
+    void write_random_data()
+    {
+        int64_t start = dsn_now_s();
+        int err = PERR_OK;
+        ASSERT_NE(pg_client, nullptr);
+
+        for (int i = 0; dsn_now_s() - start < max_detection_second; ++i %= 1000) {
+            std::string index = std::to_string(i);
+            std::string h_key = generate_hash_key_by_random(false, 0);
+            std::string s_key = "sortkey_" + index;
+            std::string value = "value_" + index;
+            err = pg_client->set(h_key, s_key, value);
+            ASSERT_EQ(err, PERR_OK);
+        }
+
+        int32_t app_id;
+        int32_t partition_count;
+        std::vector<dsn::partition_configuration> partitions;
+        ddl_client->list_app(app_name, app_id, partition_count, partitions);
+        dsn::replication::detect_hotkey_response resp;
+        dsn::replication::detect_hotkey_request req;
+        req.type = dsn::replication::hotkey_type::type::WRITE;
+        req.action = dsn::replication::detect_action::QUERY;
+        bool find_hotkey = false;
+        int partition_index;
+        for (partition_index = 0; partition_index < partitions.size(); partition_index++) {
+            req.pid = dsn::gpid(app_id, partition_index);
+            auto errinfo =
+                ddl_client->detect_hotkey(partitions[partition_index].primary, req, resp);
+            ASSERT_EQ(errinfo, dsn::ERR_OK);
+            ASSERT_EQ(resp.err_hint,
+                      "Can't get hotkey now, now state: hotkey_collector_state::STOPPED");
+        }
```

Comment:
L171 -L188 is similar with L113, maybe you can add another function to reduce duplicated code.